### PR TITLE
Add sparse block to flex_decoding kernel

### PR DESCRIPTION
**Pull request by joydddd**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #131559
* #131404
* __->__ #130884

fix typo

Finish flex_decoding block sparse

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by joydddd** [comment]
Decoding speed on noop: 1.121 (no mask) -> 1.076 (full mask 128x128 blocks) / 1.094 (empty mask)


Empty Mask: 
 Type    |   Speedup | score_mod         | dtype          | shape(B,Hq,M,Hkv,N,D)     |
|---------|-----------|-------------------|----------------|---------------------------|
| Average |     1.094 |                   |                |                           |
| Max     |     1.406 | noop_gqa          | torch.bfloat16 | (1, 16, 1, 2, 32768, 128) |
| Min     |     0.880 | relative_bias_gqa | torch.bfloat16 | (32, 16, 1, 1, 4096, 64)  |

**Comment by joydddd** [comment]
Passed IMA check.

**Comment by joydddd** [comment]
# Work around for GQA + Lacking captured buffer mask.
convert score_mod to gqa_score_mod
```python
def get_gqa_score_mod(score_mod, G, q_seq_len):
   def score_mod_gqa(score, b, hkv, m, n):
        g = m // q_seq_len
        g = torch.where(g < G, g, 0) # Apply boundary mask
        new_m = m % q_seq_len
        hq = hkv * G + g
        return score_mod(score, b, hq, new_m, n)

    return score_mod_gqa
```
convert Query to gqa format
```python
query = query.view(B, Hkv, Hq//Hkv * M, D)
```

File: test/inductor/test_flex_decoding.py [hunk 1]
```diff
@@ -38,10 +39,40 @@
...
+        g = m // q_seq_len
+        g = torch.where(g < G, g, 0)
+        new_m = m % q_seq_len
+        hq = hkv * G + g
+        return score_mod(score, b, hq, new_m, n)
+
+    return score_mod_gqa
+
+
+def get_gqa_mask_fn(mask_fn, G, q_seq_len):
```
Comment (drisspg) [inline test/inductor/test_flex_decoding.py:54]: nit: we renamed to mask_mod

File: test/inductor/test_flex_decoding.py [hunk 2]
```diff
@@ -38,10 +39,40 @@
...
+
 def create_attention(score_mod, block_mask):
     return functools.partial(flex_attention, score_mod=score_mod, block_mask=block_mask)
 
 
+def create_block_mask_test(score_mod, query, key):
+    block_mask = create_block_mask(
+        score_mod, 1, 1, query.shape[-2], key.shape[-2], query.device
+    )
+    print(block_mask.as_tuple())
```
Comment (drisspg) [inline test/inductor/test_flex_decoding.py:72]: do we want to be printing?

File: torch/_inductor/kernel/flex_decoding.py [hunk 3]
```diff
@@ -90,12 +95,39 @@ def flex_decoding_grid(batch_size, num_heads, n_keys, d_model, meta):
...
     TILE_KV = tl.cdiv(TILE_KV_OG, BLOCK_N) * BLOCK_N
 
+    TILE_KV_MULTIPLE: tl.constexpr = (TILE_KV // BLOCK_N)
+
+    SPARSE_Z = {{size("KV_NUM_BLKS", 0)}}
+    SPARSE_H = {{size("KV_NUM_BLKS", 1)}}
+
+    tl.static_assert(SPARSE_KV_BLOCK_SIZE >= BLOCK_N and SPARSE_KV_BLOCK_SIZE % BLOCK_N == 0)
+    SPARSE_KV_MULTIPLE: tl.constexpr = (SPARSE_KV_BLOCK_SIZE // BLOCK_N)
+    SPARSE_KV_BLOCK_CNT: tl.constexpr = KV_LEN // SPARSE_KV_BLOCK_SIZE
```
Comment (drisspg) [inline torch/_inductor/kernel/flex_decoding.py:105]: when we update to not even multiples will probably want tl.ceil_div() we could update now and then

File: torch/_inductor/kernel/flex_decoding.py [hunk 4]
```diff
@@ -90,12 +95,39 @@ def flex_decoding_grid(batch_size, num_heads, n_keys, d_model, meta):
...
 
     off_z = tl.program_id(0) // H
     off_h = tl.program_id(0) % H
     off_t = tl.program_id(1)
-    off_n = off_t * TILE_KV
+
+    sparse_idx_z = off_z % SPARSE_Z
+    sparse_idx_h = off_h % SPARSE_H
+
+    sparse_hz_offset = sparse_idx_z * SPARSE_H + sparse_idx_h
```
Comment (drisspg) [inline torch/_inductor/kernel/flex_decoding.py:116]: I feel like we should  be using strides instead of the size of the dim of the, that beings said I think its okay to do this, as long as we assert they are contiguous(they should be). I know Horace added some code around this in the refactor PR

File: torch/_inductor/kernel/flex_decoding.py [hunk 5]
```diff
@@ -90,12 +95,39 @@ def flex_decoding_grid(batch_size, num_heads, n_keys, d_model, meta):
...
+
+
+    block_n_start = off_t * TILE_KV_MULTIPLE                       # n_offset inside sparse block
+    block_n_end = block_n_start + TILE_KV_MULTIPLE                 # end BLOCK_N
+    block_n_last_valid = sparse_block_num * SPARSE_KV_MULTIPLE     # last valid block according to sparse mask
+    block_n_end = block_n_end if block_n_end <= block_n_last_valid else block_n_last_valid
+
+
+    indices_idx = block_n_start // SPARSE_KV_MULTIPLE
+    off_n_block_in_spase = block_n_start % SPARSE_KV_MULTIPLE
```
Comment (drisspg) [inline torch/_inductor/kernel/flex_decoding.py:128]: nit: sparse?

**Review by drisspg — APPROVED** [review]
Great clean code!

File: torch/_inductor/kernel/flex_decoding.py (hunk 4 above)
Comment (joydddd) [reply torch/_inductor/kernel/flex_decoding.py:116]: It's not in the refactor yet, so I'll skip for this PR. With that said, I think it's neccessary  for things like, e.g. packing multiple layers of precomputed masks.

File: torch/_inductor/kernel/flex_decoding.py (hunk 3 above)
Comment (joydddd) [reply torch/_inductor/kernel/flex_decoding.py:105]: Updated
